Db branching docs (#29)
metalbear-co/docs: using-mirrord/db-branching.md

---

The `db_branches` feature in mirrord lets developers spin up an isolated DB branch that mirrors the remote DB, while running safely in isolation. This allows schema changes, migrations, and experiments without impacting teammates or shared environments.
Currently, the feature is limited to MySQL databases and does not support schema or data replication.

**When is this useful?**

1. **Running schema migrations safely**  
   Developers can test migrations and schema changes without risking corruption of the remote databases.

2. **Experimenting locally**  
   Developers get a branch of the remote DB connected to their local app automatically - no admin setup required.

3. **Collaborating with teammates**  
   If a branch already exists (with the same ID), mirrord reuses it. Developers can share branches as needed.

   --- 

## Prerequisites

Before you start, make sure you have:  
1. Minimum versions installed: Operator `3.12.0`, mirrord CLI `3.160.0` and operator Helm chart `1.37.0` with `operator.mysqlBranching` value set to `true`.
2. Your local application is using environment variables to store DB connection strings.  
3. mirrord installed and working.  

> ## Configuring `mirrord db branches`

Reviewer: ```suggestion
## Configuring mirrord DB branches
```
or

```suggestion
## Configuring `db_branches`
```

or whatever, but it doesn't make sense to format `mirrord db branches` as code.

Developers define branches in their `mirrord.json`:
```json
{
  "db_branches": [
    {
      "id": "users-mysql-db", // Optional
      "type": "mysql",
      "version": "8.0",
      "name": "users-database-name",   // Optional
      "ttl_secs": 60,                 // Optional
      "connection": {
        "url": { 
            "type": "env", 
            "variable": "DB_CONNECTION_URL"  // Required
        }
      }
    }
  ]
}
```
Key Fields

> 1. id: When reused, mirrord reattaches to the same branch as long as the time-to-live (TTL) has not expired. This allows multiple sessions to share the same database branch. To prevent accidental reuse of another user’s branch, it is recommended to assign a unique value (for example, a UUID) as the identifier

Reviewer: Optional, up to you: we could format the field names as code, for example:
```suggestion
1. `id`: When reused, mirrord reattaches to the same branch as long as the time-to-live (TTL) has not expired. This allows multiple sessions to share the same database branch. To prevent accidental reuse of another user’s branch, it is recommended to assign a unique value (for example, a UUID) as the identifier
```

2. type: Currently only "mysql" is supported.
3. version: Database engine version.

> 4. name: Remote database name to clone, the override URL uses 'name' so the connection URL looks like .../dbname.

Reviewer: did you mean
```suggestion
4. name: Remote database name to clone, the override URL uses `name` so the connection URL looks like .../dbname.
```

---
title: "DB Branching"
description: "How to use mirrord to spin up isolated DB branch for safe development and testing DB migrations"
date: 2025-08-31T00:00:00+03:00
lastmod: 2025-08-31T00:00:00+03:00
draft: false
menu:
  docs:
    parent: "using-mirrord"
toc: true
tags: ["team", "enterprise"]
---

The `db_branches` feature in mirrord lets developers spin up isolated DB branch that mirror the remote DB, while running safely in isolation. This allows schema changes, migrations, and experiments without impacting teammates or shared environments.
Currently, the feature is limited to MySQL databases and does not support schema or data replication.

**When is this useful?**

1. **Running schema migrations safely**  
   Developers can test migrations and schema changes without risking corruption of the remote databases.

2. **Experimenting locally**  
   Developers get a branch of the remote DB connected to their local app automatically - no admin setup required.

3. **Collaborating with teammates**  
   If a branch already exists (with the same ID), mirrord reuses it. Developers can share branches as needed.

   --- 

## Prerequisites

Before you start, make sure you have:  

> 1. A MySQL database configured in your remote environment.  

Author: wdym? the branch you are creating is MySql db but it doesnt require an existing one?

> 1. A MySQL database configured in your remote environment.  

Reviewer: Yeah, we don't need a DB in the remote environment, since we're not copying schemas or anything like that...

2. Your local application is using environment variables to store DB connection strings.  
3. mirrord installed and working.  

## Configuring `mirrord db branches`
Developers define branches in their `mirrord.json`:
```json
{
  "db_branches": [
    {
      "id": "users-mysql-db", // Optional
      "type": "mysql",
      "version": "8.0",
      "name": "users-database-name",   // Optional
      "ttl_secs": 60,                 // Optional
      "connection": {
        "url": { 
            "type": "env", 
            "variable": "DB_CONNECTION_URL"  // Required
        }
      }
    }
  ]
}
```
Key Fields
1. id: If reused, mirrord reattaches to the same branch. Can be used sharing database branch.
2. type: Currently only "mysql" is supported.
3. version: Database engine version.
4. name: Remote database name to clone, the override URL uses 'name' so the connection URL looks like .../dbname.
If name is dropped, override URL points just to the MySQL server; app must select DB manually.

> 5. ttl_secs: Override for branch TTL (default is 30 minutes).

Author: what happends if the user set the ttl for longer than max?

6. connection.url: Required. The environment variable that contains your DB connection string.

## Running With DB Branches

> 1. Run your app 'mirrord exec' with mirrord.json configured as defined above:

Reviewer: ```suggestion
1. Run your app with mirrord and set the `db_branches` field in [the mirrord configuration file](https://metalbear.com/mirrord/docs/config).
```

That sentence is doesn't really work grammatically, and also users can run with the IDE, not only the CLI, `mirrord exec` is too specific. The colon at the end doesn't fit there, unless an introduction of a something comes before of it, and the other bullets are not in the same hierarchy level as this sentence.

> 2. mirrord will spin up a new MySQL branch (or reuse an existing one if id matches).

Reviewer: ```suggestion
2. mirrord will spin up a new MySQL branch (or reuse an existing one if the id matches).
```

    mirrord doesn’t use the original DB data. It creates a new, empty DB.

> 3. Override your DB environment variable with the branch's connection URL, so the app connects to the branch, not the source db.

Reviewer: This reads like an instruction, but isn't one. How about:

```suggestion
3. mirrord will override your DB environment variable with the branch's connection URL, so the app connects to the branch, not the source db.
```

>     This design is a safety feature - by always pointing your app to a fresh branch, it ensures that no accidental writes can ever reach the source database.

Reviewer: IDK, it's also just necessary for the feature to work at all. And accidental writes could happen if the user's app doesn't use the env var like the user thinks.

> 4. Destroy the branch automatically when TTL expires or after inactivity.

Reviewer: Also this sounds like an instruction.

```suggestion
4. The branch will be destroyed automatically when the TTL is reached while the branch is not in use (reconnecting to the same branch again extends its lifetime).
```

> 5. If an existing branch is reused, mirrord notifies you:

Reviewer: This ordering is about weird, maybe include this in or after point 2?

```

> A branch with this ID already exists for the target database.

Author: the purpose is to let the user know he's about to use an existing branch and its not in an isolated one. i want him to know that on the spot.  I agree that for how to create a fresh branch we can guide the user to the docs.
